src/openms/source/FILTERING/DATAREDUCTION/Deisotoper.cpp

```diff
@@ -38,9 +38,341 @@
+  const float averagine_check_threshold[7] = {0.0f, 0.0f, 0.05f, 0.1f, 0.2f, 0.4f, 0.6f};
+
+  // precalculate values needed for approximation of isotope distribution
+  // (lambda^k / k!)
+  std::vector<double> precalculated(max_isopeaks, 1.0);
+  {
+    // values of k!
+    UInt curr_factorial = 1;
+
+    // values of lambda ^ k
+    double curr_power = 1.0;
+
+    // lambda, value taken from Bellew et al
+    double lambda = 1.0 / 1800.0;
+    
+    for (UInt i = 1; i < max_isopeaks; ++i)
+    {
+      curr_factorial *= i;
+      curr_power *= lambda;
+      precalculated[i] = curr_power / curr_factorial;
+    }
+  }
+
+  bool has_precursor_data(false);
+  double precursor_mass(0);
+  if (old_spectrum.getPrecursors().size() == 1)
+  {
+    has_precursor_data = true;
+    int precursor_charge = old_spectrum.getPrecursors()[0].getCharge();
+    precursor_mass = (old_spectrum.getPrecursors()[0].getMZ() * precursor_charge) - (Constants::PROTON_MASS_U * precursor_charge);
+  }
+
+  for (size_t current_peak = 0; current_peak != old_spectrum.size(); ++current_peak)
+  {
+    // only process peaks which are not already in a cluster. Would form clusters identical to the cluster they are assigned to,
+    // excluding its first peak(s), because only peaks with higher mz are considered for cluster-formation.
+    if (features[current_peak] != -1) continue;
+
+    const double current_mz = old_spectrum[current_peak].getMZ();
+    const double current_intensity = old_spectrum[current_peak].getIntensity();
+    const double tolerance_dalton = fragment_unit_ppm ? Math::ppmToMass(fragment_tolerance, current_mz) : fragment_tolerance;
+
+    clusters.clear();
+    charges_of_extensions.clear();
+
+    for (int q = min_charge; q <= max_charge; ++q) // all charges are always considered -> order does not matter
+    {      
+      // try to extend isotopes from mono-isotopic peak
+
+      // do not bother testing charges q (and masses m) with: m/q > precursor_mass/q (or m > precursor_mass)
+      if (has_precursor_data)
+      {
+        double current_theo_mass = (current_mz * q) - (Constants::PROTON_MASS_U * q);
+        if (current_theo_mass > (precursor_mass + tolerance_dalton))
+        {
+          continue;
+        }
+      }
+
+      // fail early if you do not find a single extension
+      const double expected_first_mz = current_mz + Constants::C13C12_MASSDIFF_U / static_cast<double>(q);
+      Int p = old_spectrum.findNearest(expected_first_mz, tolerance_dalton);
+      if (p == -1) // test for missing peak
+      { 
+        continue;
+      }
+
+      bool found_first_extension = true;
+
+      bool has_min_isopeaks = true;
+
+      extensions.clear();
+      extensions.push_back(current_peak);
+
+      // Save frequently used values for performance reasons
+      std::vector<double> extensions_intensities = {current_intensity};
+      // This is essentially extensions.size() (which is needed a lot)
+      Int num_considered_peaks = 1;
```

Comment:
why not use `extension.size()` then?
The two variables are not always in sync below, which makes it a bit hard to see if the code is actually correct.

Reply:
Somehow thought this would improve runtime but probably does not as .size() is constant. I'll find a better solution